### ISSUE TITLE
[camerax] Level 1 Android devices can't do image analysis, video capture, and show preview at the same time

### ISSUE THREAD

**Issue by Moncader** [flutter/flutter#149947]
### Steps to reproduce

1. Make a flutter app with the attached code, and add the camera package
2. Start the app. On a level 3 device, you will probably see the counter on the bottom left rising. On a level 1 device, it will stay at 0.

### Expected results

The device should do image analysis.

### Actual results

The device doesn't do image analysis.

The reason for this was introduced with this PR: https://github.com/flutter/packages/commit/79d733cf17714244e55c673eca57850af3b7ccc6

The documentation says:
```
* **and** the camera device is at least supported hardware [`LEVEL_3`][7]
```

However the implementation isn't doing a 'least', it's checking equality.
https://github.com/flutter/packages/commit/79d733cf17714244e55c673eca57850af3b7ccc6#diff-ab4e3a5919ae7d66e10ab0ae28fa47cae44d20c78b41cbbbe05765c49e47e151R925-R928

The legacy check above that also probably has the same issue.

So basically I believe all highish-end or newer devices (like my Pixel 6 Pro that this is happening on) can't record and do analysis at the same time with the most recent version of androidx camera.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:camera/camera.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  CameraController? _controller;

  int _counter = 0;

  @override
  void initState() {
    super.initState();
    _initializeCamera();
  }

  Future<void> _initializeCamera() async {
    final cameras = await availableCameras();
    final firstCamera = cameras.first;
    _controller = CameraController(firstCamera, ResolutionPreset.medium);
    await _controller!.initialize();
    if (mounted) {
      setState(() {});
    }

    _controller!.startVideoRecording(onAvailable: _onImage);
  }

  void _onImage(CameraImage image) {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Stack(
        children: [
          Padding(
            padding: const EdgeInsets.all(32.0),
            child: Center(
              child: _controller != null
                  ? CameraPreview(_controller!)
                  : const CircularProgressIndicator(),
            ),
          ),
          Positioned(
            bottom: 16,
            right: 16,
            child: Text('images received: $_counter'),
          ),
        ],
      ),
    );
  }
}

```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
flutter doctor -v
[✓] Flutter (Channel stable, 3.22.1, on macOS 13.6.3 22G436 darwin-arm64, locale
    en-CA)
    • Flutter version 3.22.1 on channel stable at
      XXXX
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a14f74ff3a (2 weeks ago), 2024-05-22 11:08:21 -0500
    • Engine revision 55eae6864b
    • Dart version 3.4.1
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at XXXX/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = XXXX/Library/Android/sdk
    • ANDROID_SDK_ROOT = XXXX/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode-15.0.1.app/Contents/Developer
    • Build 15A507
    ✗ Unable to get list of installed Simulator runtimes.
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.90.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.91.20240529

[✓] Connected device (4 available)
    • Pixel 6 Pro (mobile)            • 1A131FDEE00480        • android-arm64  •
      Android 14 (API 34)
    • macOS (desktop)                 • macos                 • darwin-arm64   •
      macOS 13.6.3 22G436 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         •
      macOS 13.6.3 22G436 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript •
      Google Chrome 125.0.6422.142

[✓] Network resources
    • All expected network resources are available.
```
</details>

**Comment by huycozy** [comment]
I checked this on my Pixel 7 Android 14 and saw the same output result as 0 on the sample app. To reproduce this, I add a new file to the package example project at [lib folder](https://github.com/flutter/packages/blob/main/packages/camera/camera_android_camerax/example/lib):

<details>
<summary>Complete sample code</summary>

```dart
import 'package:flutter/material.dart';

import 'camera_controller.dart';
import 'camera_image.dart';
import 'camera_preview.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  CameraController? _controller;

  int _counter = 0;

  @override
  void initState() {
    super.initState();
    _initializeCamera();
  }

  Future<void> _initializeCamera() async {
    final cameras = await availableCameras();
    final firstCamera = cameras.first;
    _controller = CameraController(firstCamera);
    await _controller!.initialize();
    if (mounted) {
      setState(() {});
    }

    _controller!.startVideoRecording(onAvailable: _onImage);
  }

  void _onImage(CameraImage image) {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Stack(
        children: [
          Padding(
            padding: const EdgeInsets.all(32.0),
            child: Center(
              child: _controller != null
                  ? CameraPreview(_controller!)
                  : const CircularProgressIndicator(),
            ),
          ),
          Positioned(
            bottom: 16,
            right: 16,
            child: Text('images received: $_counter'),
          ),
        ],
      ),
    );
  }
}

```
</details>

<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel stable, 3.22.2, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.22.2 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (31 hours ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode15.3.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (3 available)
    • iPhone (mobile) • d9a94afe2b649fef56ba0bfeb052f0f2a7dae95e • ios            • iOS 15.8 19H370
    • macOS (desktop) • macos                                    • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)    • chrome                                   • web-javascript • Google Chrome 125.0.6422.142

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```
</details>

**Comment by yaakovschectman** [comment]
From triage: Our understanding of the levels is that they include Limited, Full, Legacy, Level 3, and External. Can you clarify what you mean by "level 1"?

**Comment by Moncader** [comment]
Thanks for looking in to this [yaakovschectman](https://github.com/yaakovschectman).

I was talking about 'Full' as the value of the constant was `0x1`.  I assumed that 0x1 as well as the word 'full' would mean better than 'level 3' as is the case with other media systems like Widevine DRM.

Upon reading the documentation of camerax,
https://developer.android.com/reference/android/hardware/camera2/CameraMetadata#INFO_SUPPORTED_HARDWARE_LEVEL_3
I do see that Level 3 is actually better than full, and therefore logically the current implementation is correct as per this documentation:
https://developer.android.com/media/camera/camerax/architecture#combine-use-cases

So as a camerax issue, this issue could perhaps be closed... But it does beg the question of how is it possible for the recommended android camera library to not support even the most basic of AR features on even a Google phone like Pixel 6 Pro...

**Comment by camsim99** [comment]
Thanks for the clarification @Moncader!

Followed up on this issue with CameraX folks and I found out that the logic in the plugin does in fact follow the limitations listed in https://developer.android.com/media/camera/camerax/architecture#combine-use-cases as discussed, but this is apparently out of date. There has been an update to allow concurrent image analysis, preview, and video--something to do with shared streams. This feature was added in an earlier version of CameraX than the one we currently use. Therefore, we should try to remove the logic preventing previously unsupported concurrent use cases from being used.

Closing this issue as there's nothing actionable here and filed https://github.com/flutter/flutter/issues/150387 to follow up on these new findings.
